File: apps/webflux/chatting-service/src/main/java/com/booster/chattingservice/handler/ChatWebSocketHandler.java
package com.booster.chattingservice.handler;

import com.booster.chattingservice.dto.ChatMessage;
import com.booster.chattingservice.service.ChatService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.socket.WebSocketHandler;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.WebSocketSession;
import org.springframework.web.util.UriComponentsBuilder;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import tools.jackson.databind.ObjectMapper;

import java.net.URI;

@Slf4j
@Component
@RequiredArgsConstructor
public class ChatWebSocketHandler implements WebSocketHandler {

    private final ChatService chatService;
    private final ObjectMapper objectMapper;

    @Override
    public Mono<Void> handle(WebSocketSession session) {
        // 1. Query Parameter에서 userId 추출 (예: ws://localhost:8080/ws/chat?userId=user1)
        String userId = getUserId(session);
        if (userId == null) {
            return session.close(); // ID 없으면 컷
        }

        log.info("New connection: {}", userId);

        // 2. [Input] 클라이언트 -> 서버 (Receive)
        Mono<Void> input = session.receive()
                .map(WebSocketMessage::getPayloadAsText)
                .flatMap(payload -> {
                    try {
                        // JSON 파싱 (Zero-copy는 아니지만 최대한 가볍게)
                        ChatMessage originMessage = objectMapper.readValue(payload, ChatMessage.class);
                        ChatMessage validatedMessage = originMessage.withUserId(userId);
                        return chatService.handleMessage(validatedMessage);
                    } catch (Exception e) {
                        log.error("Invalid JSON from {}: {}", userId, payload);
                        return Mono.empty(); // 에러 나도 연결은 끊지 않음 (Robustness)
                    }
                })
                .doOnTerminate(() -> {
                    // 연결 끊기면 자원 정리
                    log.info("Connection closed: {}", userId);
                    chatService.remove(userId);
                })
                .then();

        // 3. [Output] 서버 -> 클라이언트 (Send)
        // ChatService에서 나오는 Flux<ChatMessage>를 WebSocketMessage로 변환해서 전송
        Flux<WebSocketMessage> outputFlux = chatService.register(userId)
                .map(message -> {
                    try {
                        String json = objectMapper.writeValueAsString(message);
                        return session.textMessage(json);
                    } catch (Exception e) {
                        log.error("JSON Error", e);
                        return session.textMessage("Error"); // 혹은 빈 메시지
                    }
                });

        // 4. 입력과 출력을 Zip해서 세션 유지 (둘 중 하나라도 끝나면 종료됨)
        return Mono.zip(input, session.send(outputFlux)).then();
    }

    // URL에서 userId 파싱하는 헬퍼 메소드
    private String getUserId(WebSocketSession session) {
        URI uri = session.getHandshakeInfo().getUri();
        return UriComponentsBuilder.fromUri(uri)
                .build()
                .getQueryParams()
                .getFirst("userId");
    }
}

File: apps/kotlin/chatting-service/src/main/kotlin/com/booster/kotlin/chattingservice/web/ChatWebSocketHandler.kt
package com.booster.kotlin.chattingservice.web

import com.booster.kotlin.chattingservice.application.ChatService
import com.booster.kotlin.chattingservice.domain.ChatMessage
import com.booster.kotlin.core.logger
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.reactor.asFlux
import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.coroutines.reactor.mono
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.WebSocketHandler
import org.springframework.web.reactive.socket.WebSocketMessage
import org.springframework.web.reactive.socket.WebSocketSession
import org.springframework.web.util.UriComponentsBuilder
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Duration

@Component
class ChatWebSocketHandler(
    private val chatService: ChatService,
    private val objectMapper: ObjectMapper
) : WebSocketHandler {

    private val log = logger()

    override fun handle(session: WebSocketSession): Mono<Void> {
        val userId = extractUserId(session)
            ?: return session.close().also { log.warn("[REJECT] userId 없이 접속 시도") }

        log.info("[CONNECT] userId={}", userId)

        // Coroutine 컨텍스트에서 WebSocket 세션 처리
        return mono {
            // 1. Output: Kotlin Flow → Reactor Flux 변환하여 전송
            val outputFlux = chatService.register(userId)
                .map { message -> session.textMessage(objectMapper.writeValueAsString(message)) }
                .asFlux()

            // 2. Heartbeat: 30초마다 WebSocket 프로토콜 레벨 Ping 프레임 전송
            //    브라우저가 자동으로 Pong 응답 → 연결 유지
            //    클라이언트가 사라지면 TCP 재전송 실패 → 세션 종료 → remove() 호출
            val pingFlux: Flux<WebSocketMessage> = Flux.interval(Duration.ofSeconds(30))
                .map { session.pingMessage { it.wrap(ByteArray(0)) } }

            // 3. Input: TEXT 프레임만 처리 (Pong 등 프로토콜 프레임 무시)
            //    flatMap + Dispatchers.Default: Netty 이벤트 루프 블로킹 없이 처리
            //    (기존 runBlocking 제거 → Netty 스레드 고갈 방지)
            val input = session.receive()
                .filter { it.type == WebSocketMessage.Type.TEXT }
                .map { it.payloadAsText }
                .flatMap { payload ->
                    mono(Dispatchers.Default) {
                        try {
                            val message = objectMapper.readValue(payload, ChatMessage::class.java)
                            chatService.handleMessage(message.withUserId(userId))
                        } catch (e: Exception) {
                            log.warn("[PARSE] 잘못된 JSON: userId={}", userId)
                        }
                    }
                }
                .doOnTerminate {
                    log.info("[CLOSE] userId={}", userId)
                    chatService.remove(userId)
                }
                .then()

            // 4. Input + Output(메시지 + Ping) Zip으로 세션 유지
            Mono.zip(input, session.send(outputFlux.mergeWith(pingFlux))).then().awaitSingleOrNull()
        }.then()
    }

    private fun extractUserId(session: WebSocketSession): String? {
        // TODO: [Phase 1 미구현] JWT 인증으로 교체 필요
        //   현재: ?userId=xxx Query Param을 그대로 신뢰 (포트폴리오 환경)
        //   목표: ?token=<JWT> → JwtProvider.extractUserId(token) → sub 클레임에서 userId 추출
        return UriComponentsBuilder.fromUri(session.handshakeInfo.uri)
            .build()
            .queryParams
            .getFirst("userId")
    }
}
